agent/internal/containers.go

```diff
@@ -212,3 +232,168 @@ func containerEnvVars(cont cproto.Container) []string {
 		fmt.Sprintf("DET_SLOT_IDS=[%s]", strings.Join(slotIds, ",")),
 	}
 }
+
+func (c *containerManager) reattachContainers(
+	ctx *actor.Context, expectedSurvivors []aproto.ContainerReattach) (
+	[]aproto.ContainerReattachAck, error) {
+	result := make([]aproto.ContainerReattachAck, 0, len(expectedSurvivors))
+
+	runningContainers, err := c.listRunningContainers(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, survivor := range expectedSurvivors {
+		containerInfo, ok := runningContainers[survivor.Container.ID]
+		var ack aproto.ContainerReattachAck
+
+		if !ok {
+			ack = aproto.ContainerReattachAck{
+				Container: cproto.Container{ID: survivor.Container.ID},
+				Failure: &aproto.ContainerFailure{
+					FailureType: aproto.AgentFailed,
+					ErrMsg:      "container is gone on reattachment",
+				},
+			}
+		} else {
+			ctx.Log().Infof("will reattach container %s", survivor.Container.ID)
+			cpc, err := c.reattachContainer(ctx, survivor.Container, containerInfo)
+			if err != nil {
+				ack = aproto.ContainerReattachAck{
+					Container: *cpc,
+					Failure: &aproto.ContainerFailure{
+						FailureType: aproto.AgentFailed,
+						ErrMsg:      "failed to restore info from container labels",
+					},
+				}
+			} else {
+				ack = aproto.ContainerReattachAck{
+					Container: cproto.Container{ID: survivor.Container.ID},
+				}
+			}
+		}
+
+		result = append(result, ack)
+		delete(runningContainers, survivor.Container.ID)
+	}
+
+	// SIGKILL the rest.
+	for cid, containerInfo := range runningContainers {
+		ctx.Log().Infof("will kill container %s", cid)
+		err := c.docker.ContainerKill(
+			context.Background(), containerInfo.ID, unix.SignalName(unix.SIGKILL))
+		if err != nil {
+			ctx.Log().WithError(err).Warnf("failed to kill container %s", cid)
+		}
+	}
+
+	return result, nil
+}
+
+func (c *containerManager) reattachContainer(
+	ctx *actor.Context, containerPrevState cproto.Container, containerInfo types.Container) (
+	*cproto.Container, error) {
+	containerCurrState, err := c.unmakeContainerDockerLabels(ctx, containerInfo)
+	if err != nil {
+		return nil, err
+	}
+	// TODO(ilia): Support reattaching containers that have changed state:
+	// - starting -> running,
+	// - running -> terminated.
+	if containerCurrState.State != containerPrevState.State {
+		return nil, errors.New("container has changed state while offline")
+	}
+
+	cid := containerPrevState.ID
+	_, ok := ctx.ActorOf(cid, reattachContainerActor(*containerCurrState, c.docker))
+	if !ok {
+		errorMsg := fmt.Sprintf("failed to reattach container %s: actor already exists", cid)
+		ctx.Log().Warnf(errorMsg)
+		if killed := ctx.Kill(cid); killed {
```

Comment:
could sigkill the survivor here too, but this is already very unlikely (just dont remove it from runningContainers in the event of certain errors?)

Reply:
oops, yep nice